fundamental-concepts/data-structures-stdlib/10_optional.md

```diff
@@ -0,0 +1,72 @@
+# Optional fields
+
+Sometimes you need to deal with data fields that are **optional**. This means that the value might
+not be specified at the time of creating the record. For example, you might want to add an optional
+email address field to the `Person` record:
+
+<pre class="file" data-target="clipboard">
+import DA.Optional
+
+data Person = Person {
+  name: Text,
+  address: Address,
+  age: Int,
+  email: Optional Text
+}
+</pre>
+
+Optional fields are either present
+
+<pre class="file" data-target="clipboard">
+testPerson: Person
+testPerson = Person {name = "Alice", address = Address {street = "RabbitStreet", city = "TheBigUnknown", country = "Wonderland"}, age = 7, email = Some "alice@wonderland.com"}
+</pre>
+
+or absent 
+
+<pre class="file" data-target="clipboard">
+testPerson: Person
+testPerson = Person {name = "Alice", address = Address {street = "RabbitStreet", city = "TheBigUnknown", country = "Wonderland"}, age = 7, email = None}
+</pre>
+
+Likewise, you can pattern match against the two constructors `Some: a -> Optional a` and `None: Optional a` to deconstruct an optional value:
+
+<pre class="file" data-filename="daml/AddressBook.daml" data-target="append">
+prettyEmail: Optional Text -> Text
+prettyEmail mbEmail = case mbEmail of
+  None -> " -- "
+  Some email -> email
+</pre>
+
+Optional values can also be used as return values of functions, if you want to indicate that the
+computation might fail or might be undefined. `lookup` of `DA.Next.Map` is an archetypal example:
+
+<pre>
+lookup: Map k v -> Optional v
+</pre>
+
+The `lookup` function makes it very clear by it's type that it might fail and there does not exist a
+value with the given key. The `Optional` return value forces the consumer to deal with the possible
+failure.
+
+Have a look at the
+[documentation](https://docs.daml.com/daml/stdlib/DA-Optional.html#module-da-optional), then try to
+answer the following quiz:
+
+>> Which functions converts an optional value to a list with possibly 0 or 1 element? <<
+=== optionalToList
+
+>> Which function of type f: a -> Optional a -> a converts an optional value to a value, given a default case for when the optional value is None ? <<
+=== fromOptional
+
+>> Convert the list of optional values 'xs: [Optional Text]' to a list with of type '[Text]' which contains only the value constructed with 'Some'. <<
+=== catOptionals xs
+
```

Comment:
```suggestion
```
Would remove this question

Reply:
Wouldn't, `catOptionals` is quite an important and often used function.

Reply:
Sure, but we can't teach them everything, some things they have to discover themselves. Biggest issue here is getting frustrated with the number of answers one needs to provide within the whole scenario (overall) as there's no option of progressing to the next step unless all the answers are correct. With previous questions on maps and lists you've really done a great job of pointing the users to the place where they can get necessary info on all things map related _when_ they decide to use it. That's the key.